### PR TITLE
SMP init

### PR DESCRIPTION
**Pull request by TezRomacH**
## Description

- Added `prefix` parameter for registry from a module. Example `r.add_from_module(module, prefix="module.")`
- Added support for [SMP](https://github.com/qubvel/segmentation_models.pytorch) models to the Config API
- hotfix IoU with no reduce dim
- Added test example for segmentation

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [Code of Conduct](https://github.com/catalyst-team/catalyst/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I have read the [Contributing](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md) guide.
- [x] I have checked the code-style using `make check-style`.
- [x] I have written the docstring in Google format for all the methods and classes that I used.
- [x] I have checked the docs using `make check-docs`.

### CODE REVIEW

**Comment by Scitator** [comment]
@TezRomacH 
Hi, could you please merge with master and add some tests for modules import and their Config API usage?

File: catalyst/contrib/transforms/tensor.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import numpy as np
+
+from albumentations.core.transforms_interface import ImageOnlyTransform
+import torch
+
+
+class TensorToImage(ImageOnlyTransform):
```
Comment (Scitator) [inline catalyst/contrib/transforms/tensor.py:10]: why do we need this transforms here?
@bagxi what do you think?

File: catalyst/dl/registry.py [hunk 2]
```diff
@@ -24,6 +24,12 @@ def _transforms_loader(r: Registry):
     try:
         import albumentations as m
         r.add_from_module(m)
+
+        from albumentations import pytorch as p
+        r.add_from_module(p)
```
Comment (Scitator) [inline catalyst/dl/registry.py]: what to you think about adding `albu` prefix?

Comment (Scitator) [reply catalyst/dl/registry.py]: or, maybe `A` for Albumentations and `C` for Catalyst?

File: catalyst/contrib/transforms/tensor.py (hunk 1 above)
Comment (Scitator) [reply catalyst/contrib/transforms/tensor.py:10]: btw, we also need to merge this PR first
https://github.com/catalyst-team/catalyst/pull/604

Comment (TezRomacH) [reply catalyst/contrib/transforms/tensor.py:10]: This transformation is needed when torch.Tensor is returned from the dataset, not numpy. For example, MNIST in our tests

File: catalyst/dl/registry.py (hunk 2 above)
Comment (TezRomacH) [reply catalyst/dl/registry.py]: We do not support any other augmentation library (I think, we won't).

And I think it's unnecessary to add a prefix to all transforms in config

Comment (TezRomacH) [reply catalyst/dl/registry.py]: About, `C` for Catalyst. Users will not use our contrib transforms very often, I prefer to use full form, it is much readable and clean

File: catalyst/contrib/transforms/tensor.py (hunk 1 above)
Comment (Scitator) [reply catalyst/contrib/transforms/tensor.py:10]: maybe, it should be `TensorToNumpy` then?

Comment (Scitator) [reply catalyst/contrib/transforms/tensor.py:10]: or just `ToNumpy`?

File: catalyst/dl/registry.py (hunk 2 above)
Comment (Scitator) [reply catalyst/dl/registry.py]: there are 2 ways:
1. we add prefix for all our sources for transformations, except Core Catalyst one
2. we don't use any prefixes at all... but it could cause name collision
all other variants looks unscalable

PS. And yup, we will support more transformations source in the future

Comment (TezRomacH) [reply catalyst/dl/registry.py]: > we add a prefix for all our sources for transformations, except Core Catalyst one
why is that? `albumentations` has a lot more common transforms than in catalyst.core

it's totally okay to use prefixes for some unusual classes or methods. Catalyst has only two transforms, and maybe in the future, it will have a few more. but albu has a lot of common transformations

Comment (TezRomacH) [reply catalyst/dl/registry.py]: > PS. And yup, we will support more transformations source in the future

nice, and for them, i would favor adding a prefix, like `somenewlib.Transforms`

but making `albumentations` preferred one, without prefix

Comment (TezRomacH) [reply catalyst/dl/registry.py]: look at a typical transform config
https://github.com/catalyst-team/catalyst/pull/600/files#diff-e8e0a33e0d5dfbfe552d3a0fc812d1bdR5

adding an `albu.` prefix for every single transform is long and unnecessary

File: catalyst/contrib/transforms/tensor.py (hunk 1 above)
Comment (TezRomacH) [reply catalyst/contrib/transforms/tensor.py:10]: it's okay to me, but this transform was already in catalyst.

so we need @bagxi  to discuss

Comment (bagxi) [reply catalyst/contrib/transforms/tensor.py:10]: maybe, `TensorToImage`?

Comment (bagxi) [reply catalyst/contrib/transforms/tensor.py:10]: and would it be useful if we will add parameter like `use catalyst.utils.tensor_to_ndimage`?

Comment (TezRomacH) [reply catalyst/contrib/transforms/tensor.py:10]: let's make `tensor_to_ndimage` more general with flags like `need_normalization` and `need_move_channels_dim` and then use this function inside of `TensorToImage`/`ToNumpy`/`YouNameIt` transfrom

Comment (Scitator) [reply catalyst/contrib/transforms/tensor.py:10]: prosed simplification, `normalize=`, `move_channels_dim=`

File: catalyst/dl/registry.py (hunk 2 above)
Comment (Scitator) [reply catalyst/dl/registry.py]: `but making albumentations preferred one, without prefix` why?

Comment (TezRomacH) [reply catalyst/dl/registry.py]: ODS support

Comment (TezRomacH) [reply catalyst/dl/registry.py]: the main point of this PR is a friendship between Catalyst, albumentations, SMP and maybe ttach in future

File: catalyst/contrib/transforms/tensor.py (hunk 1 above)
Comment (bagxi) [reply catalyst/contrib/transforms/tensor.py:10]: maybe it is better to use `denormalize=` (instead of `normalize=`) as in `utils.tensor_to_ndimage` we are multiplying image by std and adding mean?

Comment (Scitator) [reply catalyst/contrib/transforms/tensor.py:10]: sure

File: catalyst/dl/registry.py (hunk 2 above)
Comment (Scitator) [reply catalyst/dl/registry.py]: I dont want to have any possible bias to any other framework (except Pytorch, of course),
I like ODS, nevertheless, we need to be open for any other framework on the same rules.

Comment (Scitator) [reply catalyst/dl/registry.py]: you can use any prefix, even `A` is good enough, 
or add some global prefix-source for all transformations

Comment (TezRomacH) [reply catalyst/dl/registry.py]: ok, i've added `A.` prefix

File: examples/_tests_mnist_stages2/config1.yml [hunk 3]
```diff
@@ -17,11 +17,11 @@ stages:
   transform_params:
     transform: A.Compose
     transforms:
-      - transform: catalsyt.TensorToImage
+      - transform: C.TensorToImage
       - transform: A.Normalize
         mean: [0.1307]
         std: [0.3081]
-      - transform: catalsyt.ToTensor
+      - transform: C.ToTensor
```
Comment (TezRomacH) [inline examples/_tests_mnist_stages2/config1.yml]: but why?

we have never called our framework `C`. using `catalsyt.ToTensor` is much clearer

Comment (TezRomacH) [reply examples/_tests_mnist_stages2/config1.yml]: @Scitator

Comment (Scitator) [reply examples/_tests_mnist_stages2/config1.yml]: first of all - **C**ata**lyst** :)
but, `Catalyst.ToTensor` looks too long and complicated

Comment (TezRomacH) [reply examples/_tests_mnist_stages2/config1.yml]: `from catalyst import utils`

it's lowercase :)

and here: `transform: catalsyt.ToTensor` it means to use  `ToTensor` from sort of package `catalyst.`

Comment (TezRomacH) [reply examples/_tests_mnist_stages2/config1.yml]: `C.ToTensor` is unclear and more complicated

Comment (Scitator) [reply examples/_tests_mnist_stages2/config1.yml]: why do you think so? why is it unclear and more complicated?

Comment (TezRomacH) [reply examples/_tests_mnist_stages2/config1.yml]: because, you will never know, what is `C.`, where else it's used

`smp.Unet` means you use Unet from module `segmentation_models_pytorch` also known as `smp`

`A.RandomCrop` means you use RandomCrop from `albumentations` also known as `A`

what does `C.ToTensor` mean?

Comment (TezRomacH) [reply examples/_tests_mnist_stages2/config1.yml]: To peacefully end the argue, I suggest adding functionality that allows you to add several different prefixes for the same method or class.

For example, `.add_from_module(t, prefix=["catalyst.", "C."])`.

Comment (Scitator) [reply examples/_tests_mnist_stages2/config1.yml]: you know `albumentaions as A`, nevertheless from my perspective they are `albu` or `albumentations `... it's a matter of taste.

but the idea with  different prefixes looks good, you are right :)

Comment (Scitator) [reply examples/_tests_mnist_stages2/config1.yml]: you can also add aliases for albumentations, like `albumentations`, `albu`, `A`

Comment (TezRomacH) [reply examples/_tests_mnist_stages2/config1.yml]: albumentations uses `A.` in their first and examples :)

Comment (TezRomacH) [reply examples/_tests_mnist_stages2/config1.yml]: * in their docs

Comment (Scitator) [reply examples/_tests_mnist_stages2/config1.yml]: would you like to make changes during this PR, or is it okay for you to make them on https://github.com/catalyst-team/catalyst/pull/607 ?
GANs are waiting for you, @TezRomacH :)

Comment (TezRomacH) [reply examples/_tests_mnist_stages2/config1.yml]: I am literally working on this right now

give me 30-40 minutes

Comment (TezRomacH) [reply examples/_tests_mnist_stages2/config1.yml]: done

**Comment by TezRomacH** [comment]
build is done successfully, @Scitator
